### PR TITLE
Foreign Keys

### PR DESCRIPTION
**Pull request by Hydrocharged**

### CODE REVIEW

File: go/vt/sqlparser/ast.go [hunk 1]
```diff
@@ -886,6 +886,17 @@ func (node *DDL) Format(buf *TrackedBuffer) {
 			buf.Myprintf("%s table %v %s column %v to %v", node.Action, node.Table, node.ColumnAction, node.Column, node.ToColumn)
 		} else if node.IndexSpec != nil {
 			buf.Myprintf("%s table %v %v", node.Action, node.Table, node.IndexSpec)
+		} else if node.TableSpec != nil && len(node.TableSpec.Constraints) == 1 {
+			fkDef, ok := node.TableSpec.Constraints[0].Details.(*ForeignKeyDefinition)
+			if ok {
+				if fkDef.IsDrop {
+					buf.Myprintf("%s table %v drop foreign key %s", node.Action, node.Table, node.TableSpec.Constraints[0].Name)
```
Comment (zachmu) [inline go/vt/sqlparser/ast.go:893]: We know this is ALTER right?

File: go/vt/sqlparser/ast.go [hunk 2]
```diff
@@ -1720,8 +1731,9 @@ func (a ReferenceAction) Format(buf *TrackedBuffer) {
 	}
 }
 
-// ForeignKeyDefinition describes a foreign key in a CREATE TABLE statement
+// ForeignKeyDefinition describes a foreign key
 type ForeignKeyDefinition struct {
+	IsDrop            bool
```
Comment (zachmu) [inline go/vt/sqlparser/ast.go:1736]: Think I would prefer this to be an action string, as in other DDL / column actions

File: go/vt/sqlparser/sql_test.go [hunk 3]
```diff
@@ -51,7 +51,7 @@ func TestAlterCreateIndex(t *testing.T) {
 				Columns: []*IndexColumn{
 					{Column: ColIdent{val: "bestcol"}},
 				},
-				Options: []*IndexOption{{Name: "comment", Value: &SQLVal{Type:StrVal, Val: []byte("hello world")}}},
+				Options: []*IndexOption{{Name: "COMMENT", Value: &SQLVal{Type:StrVal, Val: []byte("hello world")}}},
```
Comment (zachmu) [inline go/vt/sqlparser/sql_test.go:54]: Thanks for fixing, not sure how i missed these

File: go/vt/sqlparser/ast.go (hunk 2 above)
Comment (zachmu) [reply go/vt/sqlparser/ast.go:1736]: Thinking about this a little further, I don't think this belongs in the definition struct at all. This should be a "constraintAction" field in the parent DDL action struct. Column modifications / drops / adds have a similar structure you should copy.

File: go/vt/sqlparser/sql_test.go [hunk 4]
```diff
@@ -280,6 +280,264 @@ func TestShowIndex(t *testing.T) {
 	}
 }
 
+func TestCreateForeignKey(t *testing.T) {
+	tests := []testForeignKeyStruct{
```
Comment (zachmu) [inline go/vt/sqlparser/sql_test.go:284]: Great tests. Would add one more, for the base syntax (no constraint keyword, unnamed constraint) without any ON UPDATE ON DELETE clauses.

File: go/vt/sqlparser/sql_test.go [hunk 5]
```diff
@@ -280,6 +280,264 @@ func TestShowIndex(t *testing.T) {
...
+		},
+	}
+
+	testForeignKey(t, tests)
+}
+
+func TestAlterDropForeignKey(t *testing.T) {
+	tests := []testForeignKeyStruct{
+		{
+			`ALTER TABLE child DROP FOREIGN KEY fk_parent_id`,
```
Comment (zachmu) [inline go/vt/sqlparser/sql_test.go:509]: Should this work for unnamed foreign keys constraints too? How does MySQL do this?

File: go/vt/sqlparser/sql_test.go [hunk 6]
```diff
@@ -280,6 +280,264 @@ func TestShowIndex(t *testing.T) {
...
+		},
+	}
+
+	testForeignKey(t, tests)
+}
+
+func TestAlterAddForeignKey(t *testing.T) {
+	tests := []testForeignKeyStruct{
+		{
+			`ALTER TABLE child ADD
```
Comment (zachmu) [inline go/vt/sqlparser/sql_test.go:408]: Same comment about testing more syntax permutations. Need one that does ADD FOREIGN KEY (cols) REFERNCES ...  without any ON UPDATE or ON DELETE

File: go/vt/sqlparser/ast.go (hunk 1 above)
Comment (Hydrocharged) [reply go/vt/sqlparser/ast.go:893]: Yep

File: go/vt/sqlparser/sql_test.go (hunk 5 above)
Comment (Hydrocharged) [reply go/vt/sqlparser/sql_test.go:509]: If you don't provide a name then MySQL generates a name for you, and you'll have to use that name to drop it.

**Pull request merged**